File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Java/c9fb2d1a92fd15a81ee8f3fdd5b726d0.java
import java.util.*;
import java.io.*;

class DSum {
    static class Solver {
        private final BufferedReader in;
        private final PrintWriter out;

        public Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        private void solve() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            int n = Integer.parseInt(st.nextToken());
            int k = Integer.parseInt(st.nextToken());

            long[] sums = new long[n];
            List<List<Integer>> arrs = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                st = new StringTokenizer(in.readLine());
                int t = Integer.parseInt(st.nextToken());
                List<Integer> arr = new ArrayList<>();
                for (int j = 0; j < t; j++) {
                    arr.add(Integer.parseInt(st.nextToken()));
                }
                if (t > k) {
                    arr = arr.subList(0, k);
                }
                sums[i] = sum(arr);
                arrs.add(arr);
            }

            long[][] dps = new long[13][k + 1];
            out.println(rec(0, n, k, dps, sums, arrs));
        }

        private long sum(List<Integer> arr) {
            long sum = 0;
            for (int num : arr) {
                sum += num;
            }
            return sum;
        }

        private void doCopy(int dpi, long[][] dps) {
            System.arraycopy(dps[dpi], 0, dps[dpi + 1], 0, dps[0].length);
        }

        private void addToKnap(int i, int dpi, long[][] dps, long[] sums, List<List<Integer>> arrs) {
            long v = sums[i];
            int k1 = arrs.get(i).size();
            for (int j = k; j >= k1; j--) {
                dps[dpi][j] = Math.max(dps[dpi][j], dps[dpi][j - k1] + v);
            }
        }

        private long rec(int l, int r, int k, long[][] dps, long[] sums, List<List<Integer>> arrs) {
            long ans = 0;
            List<Integer> carr = arrs.get(l);
            long[] dp = dps[k];
            if (l + 1 == r) {
                ans = dp[k];
                long sum = 0;
                for (int i = 0; i < Math.min(k, carr.size()); i++) {
                    sum += carr.get(i);
                    ans = Math.max(ans, dp[k - i - 1] + sum);
                }
            } else {
                int m = (l + r) / 2;
                doCopy(k, dps);
                for (int i = l; i < m; i++) {
                    addToKnap(i, k + 1, dps, sums, arrs);
                }
                ans = Math.max(ans, rec(m, r, k, dps, sums, arrs));
                doCopy(k, dps);
                for (int i = m; i < r; i++) {
                    addToKnap(i, k + 1, dps, sums, arrs);
                }
                ans = Math.max(ans, rec(l, m, k, dps, sums, arrs));
            }
            return ans;
        }
    }

    public static void solve(BufferedReader in, PrintWriter out) throws IOException {
        Solver solver = new Solver(in, out);
        solver.solve();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        solve(in, out);
        out.close();
    }
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/c9fb2d1a92fd15a81ee8f3fdd5b726d0.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define max(a, b) ((a) > (b) ? (a) : (b))

typedef long long ll;
typedef struct {
    int size;
    int* arr;
} vi;
typedef struct {
    int size;
    vi* arrs;
} vvi;
typedef struct {
    int size;
    ll* arr;
} vll;
typedef struct {
    int size;
    vll* arrs;
} vvll;

int k;
vvll dps;
vll sums;
vvi arrs;

void do_copy(int dpi) {
    memcpy(dps.arrs[dpi + 1].arr, dps.arrs[dpi].arr, sizeof(dps.arrs[0].arr[0]) * (k + 1));
}

void add_to_knap(int i, int dpi) {
    ll v = sums.arr[i];
    int k1 = arrs.arrs[i].size;
    for (int j = k; j >= k1; j--) {
        dps.arrs[dpi].arr[j] = max(dps.arrs[dpi].arr[j], dps.arrs[dpi].arr[j - k1] + v);
    }
}

ll rec(int l, int r, int dpi) {
    ll ans = 0;
    vi carr = arrs.arrs[l];
    vll dp = dps.arrs[dpi];
    if (l + 1 == r) {
        ans = max(ans, dp.arr[k]);
        ll sum = 0;
        for (int i = 0; i < min(k, carr.size); i++) {
            sum += carr.arr[i];
            ans = max(ans, dp.arr[k - i - 1] + sum);
        }
    } else {
        int m = (l + r) / 2;
        do_copy(dpi);
        for (int i = l; i < m; i++) {
            add_to_knap(i, dpi + 1);
        }
        ans = max(ans, rec(m, r, dpi + 1));
        do_copy(dpi);
        for (int i = m; i < r; i++) {
            add_to_knap(i, dpi + 1);
        }
        ans = max(ans, rec(l, m, dpi + 1));
    }
    return ans;
}

void solve() {
    int n;
    scanf("%d %d", &n, &k);
    sums.size = n;
    sums.arr = (ll*)malloc(sizeof(ll) * n);
    arrs.size = n;
    arrs.arrs = (vi*)malloc(sizeof(vi) * n);
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        arrs.arrs[i].size = t;
        arrs.arrs[i].arr = (int*)malloc(sizeof(int) * t);
        for (int j = 0; j < t; j++) {
            scanf("%d", &arrs.arrs[i].arr[j]);
        }
        if (t > k) {
            arrs.arrs[i].size = k;
        }
        sums.arr[i] = 0;
        for (int j = 0; j < arrs.arrs[i].size; j++) {
            sums.arr[i] += arrs.arrs[i].arr[j];
        }
    }
    dps.size = 13;
    dps.arrs = (vll*)malloc(sizeof(vll) * 13);
    for (int i = 0; i < 13; i++) {
        dps.arrs[i].size = k + 1;
        dps.arrs[i].arr = (ll*)malloc(sizeof(ll) * (k + 1));
    }
    printf("%lld\n", rec(0, n, 0));
}

int main() {
    solve();
    return 0;
}
